content/blog/2019-11-13-DynamicEdgeProfiling/index.md

```diff
@@ -0,0 +1,122 @@
+
+As the profiling logic and data are written in a runtime library disjoint from the LLVM pass, this information can be outputted however the user wants with minor edits to the library file. We chose to print it to standard out for simplicity. There is also the matter of correlating the block numbers assigned during the LLVM pass to the actual blocks, and this is done by printing the block-to-number matching for the source and destination blocks of each instrumented edge to standard out during the pass.
+
+## Evaluation
+### Correctness
+To measure performance, we ran a subset of C benchmarks from the [LLVM test suite](https://llvm.org/docs/TestSuiteGuide.html). We collected the tabulated results and CFGs to verify our instrumentation was on an optimal number of edges. We then ran the algorithm described above to extrapolate the uninstrumented edges. We include one test here for visibility.
+Consider the following CFGS for a program with two functions, main and testFunc.
+ <!-- ![](https://i.imgur.com/h9Stzm2.png =300x) ![](https://i.imgur.com/ls7PM8e.png =300x)  -->
+<img src="main.png"/>
+<img src="testfunc.png"/>
+We run our instrumented program with input argument `5` and procure the table below.
+
+| Edge     | Count |
+|----------|-------|
+| [6->4]   | 5     |
+| [9->10]  | 1     |
+
+When we ran our pass, we also output the edges that were instrumented.
+These edges were `[0->2]`, `[6->4]`, and `[9->10]`.
+
+To verify these results, we consider our program. 
+```
+void testFunc(int n){
+    if(n%2==1) printf("The number is odd\n");
+}
+
+int main(int num) {
+    for(int i =0; i<5; i++){
+        printf("Iteration %i\n", i);
+    }
+    if(num%2 == 0){
+        printf("The number is even\n");
+    }
+    else{
+        testFunc(num);
+    }
+    printf("%i\n", num + 2);
+    return 0;
+}
+```
+In order to automatically verify these results, we implement a naive LLVM pass for edge profiling that instruments every edge. We then compare the results collected and extrapolated from optimal edge profiling with the results from running the naive implementation.
```

Comment:
I notice that your earlier instance of "naïve" had an umlaut. Any reason these later occurrences were left out? :smiley: